### ISSUE TITLE
Question on data parallel: dose the training code support data parallel with fsdp?

### ISSUE THREAD

**Issue by hust-umi** [nvidia-cosmos/cosmos-predict2.5#143]
dose the training code support data parallel with megatron? 

I know that context parallel has already been supported. I wondering that is data parallel also supported? As I found no  gradient sync operation in the trainer loop in ``cosmos_predict2/_src/imaginaire/trainer.py``

**Comment by hust-umi** [comment]
I did find a bug on data parallel with fsdp when ``trainer.distributed_parallelism==fsdp and model.config.fsdp_shard_size==1`` holds.
The problem is: ``trainer.distributed_parallelism==fsdp`` avoids ddp wrapping in ``cosmos_predict2/_src/imaginaire/trainer.py:Trainer.train`` and ``model.config.fsdp_shard_size==1`` avoids ``cosmos_predict2/_src/predict2/utils/model_loader.py:create_model_from_consolidated_checkpoint_with_fsdp`` to wrap the model with fsdp. Without any parallelism wrapping to the model, gradients can not sync automatically and affects training.

**Comment by hust-umi** [comment]
Exactly, I have check the gradient in ``ImaginaireTrainer.training_step`` with fsdp enabled and fsdp shard size=1 with the following function:
```python
def check_ddp_gradients_with_dtensor(
    model: torch.nn.Module,
    iteration: int | None = None,
    only_rank0_print: bool = True,
    atol: float = 1e-6,
    rtol: float = 1e-5,
    check_nonfinite: bool = False,
) -> dict:
    if hasattr(model, "module"):
        model_to_check = model.module
    else:
        model_to_check = model

    if not dist.is_available() or not dist.is_initialized():
        return {
            "all_equal": True,
            "mismatched_params": [],
            "missing_grad_params": [],
            "nonfinite_params": [],
        }

    rank = dist.get_rank()
    world_size = dist.get_world_size()

    mismatched_params = []
    missing_grad_params = []
    nonfinite_params = []

    for name, param in model_to_check.named_parameters():
        if param.grad is None:
            missing_grad_params.append(name)
            continue

        grad_local, grad_type = _unwrap_grad_for_debug(param.grad)

        if check_nonfinite and not torch.isfinite(grad_local).all():
            nonfinite_params.append(name)

        stats = torch.tensor(
            [
                grad_local.sum().item(),
                grad_local.mean().item(),
                grad_local.norm().item(),
                grad_local.abs().max().item(),
                float(grad_local.numel()),
            ],
            device=grad_local.device,
            dtype=torch.float32,
        )

        gathered = [torch.zeros_like(stats) for _ in range(world_size)]
        dist.all_gather(gathered, stats)

        ref = gathered[0]
        same = True
        for i in range(1, world_size):
            if not torch.allclose(gathered[i], ref, atol=atol, rtol=rtol):
                same = False
                break

        if not same and grad_type == "tensor":
            mismatched_params.append(name)

    all_equal = len(mismatched_params) == 0 and len(nonfinite_params) == 0

    should_print = (rank == 0) if only_rank0_print else True
    if should_print:
        it_str = f"[iter={iteration}] " if iteration is not None else ""
        log.critical(
            f"{it_str}DDP/DTensor grad debug check: "
            f"all_equal={all_equal}, "
            f"mismatched={len(mismatched_params)}, "
            f"missing_grad={len(missing_grad_params)}, "
            f"nonfinite={len(nonfinite_params)}"
        )

    return {
        "all_equal": all_equal,
        "mismatched_params": mismatched_params,
        "missing_grad_params": missing_grad_params,
        "nonfinite_params": nonfinite_params,
    }
```

and change ``if fsdp_shard_size > 1:`` to ``if config.trainer.distributed_parallelism == "fsdp":`` in ``create_model_from_consolidated_checkpoint_with_fsdp`` would avoid the problem.
